core/src/main/scala/kafka/server/Auditor.scala

```diff
@@ -0,0 +1,52 @@
+/**
+  * Licensed to the Apache Software Foundation (ASF) under one or more
+  * contributor license agreements.  See the NOTICE file distributed with
+  * this work for additional information regarding copyright ownership.
+  * The ASF licenses this file to You under the Apache License, Version 2.0
+  * (the "License"); you may not use this file except in compliance with
+  * the License.  You may obtain a copy of the License at
+  *
+  *    http://www.apache.org/licenses/LICENSE-2.0
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+package kafka.server
+
+import java.util.concurrent.TimeUnit
+import kafka.network.RequestChannel
+import org.apache.kafka.common.requests.AbstractResponse
+import org.apache.kafka.common.Configurable
+
+/**
+  * Top level interface that all pluggable auditor must implement. Kafka will read the 'auditor.class.name' config
+  * value at startup time, create an instance of the specificed class using the default constructor, and call its
+  * 'configure' method.
+  *
+  * From that point onwards, every pair of request and response will be routed to the 'record' method.
+  *
+  * If 'auditor.class.name' has no value specified or the specified class does not exist, the <code>NoOpAuditor</code>
+  * will be used as a place holder.
+  */
+trait Auditor extends Configurable {
```

Comment:
Does this need to be scala code?

Reply:
1. I defined the auditor interface in the "kafka.server" package since I think it is the most reasonable directory to define it. This package is under "core/src/main/**scala**/kafka/server", it should be written in scala.

2. This broker-side auditor is similar to the authorizer which also has its interface defined in Scala and some non-trivial implementations defined in Java in kafka-server.

3. I can't think of any benefit if the interface is defined in Java since some conversion from Scala data type to Java data type needs to happen anyway. Let me know what I missed.

Reply:
i'd call this "Observer" (as its the more general purpose, auditing being a special case).
I agree about the scala - to be consistent with Authorizer if nothing else.

Reply:
Good idea!